File: src/lib/puzzle/generator.ts
import type { GridSize, Difficulty, Puzzle } from '../../types/puzzle';
import { solveWithRandomization, hasUniqueSolution } from './solver';
import { getTargetClueCount, assessDifficulty } from './difficulty';
import { getGridConfig } from '../../constants/grid-configs';

function generateId(): string {
  return Math.random().toString(36).substring(2, 10);
}

function shuffle<T>(array: T[]): T[] {
  const result = [...array];
  for (let i = result.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [result[i], result[j]] = [result[j], result[i]];
  }
  return result;
}

export function generatePuzzle(
  size: GridSize,
  difficulty: Difficulty
): Puzzle {
  // For large grids (16x16, 25x25), use fast pattern-based generation
  if (size >= 16) {
    return generateLargeGridPuzzle(size, difficulty);
  }

  // For smaller grids, use backtracking with unique solution check
  const emptyGrid: (number | null)[] = new Array(size * size).fill(null);
  const result = solveWithRandomization(emptyGrid, size);

  if (!result.solved || !result.solution) {
    throw new Error('Failed to generate complete grid');
  }

  const solution = result.solution;
  const targetClues = getTargetClueCount(size, difficulty);
  const puzzle = removeCellsWithCheck(solution, size, targetClues);
  const analysis = assessDifficulty(puzzle, size);

  return {
    id: generateId(),
    size,
    difficulty: analysis.difficulty,
    cells: puzzle,
    solution,
  };
}

// Fast generation for large grids using pattern shuffling
function generateLargeGridPuzzle(
  size: GridSize,
  difficulty: Difficulty
): Puzzle {
  const config = getGridConfig(size);
  const boxRows = config.boxConfig.rows;
  const boxCols = config.boxConfig.cols;

  // Step 1: Create a valid base grid using a simple pattern
  const solution = createPatternGrid(size, boxRows, boxCols);

  // Step 2: Shuffle to randomize
  shuffleGrid(solution, size, boxRows, boxCols);

  // Step 3: Remove cells (no unique solution check for speed)
  const targetClues = getTargetClueCount(size, difficulty);
  const puzzle = removeCellsSimple(solution, targetClues);

  return {
    id: generateId(),
    size,
    difficulty,
    cells: puzzle,
    solution,
  };
}

// Creates a valid Sudoku grid using a mathematical pattern
function createPatternGrid(size: number, boxRows: number, boxCols: number): number[] {
  const grid: number[] = new Array(size * size);

  for (let row = 0; row < size; row++) {
    for (let col = 0; col < size; col++) {
      // This formula creates a valid Sudoku pattern
      const value = (boxCols * (row % boxRows) + Math.floor(row / boxRows) + col) % size + 1;
      grid[row * size + col] = value;
    }
  }

  return grid;
}

// Shuffle the grid while maintaining validity
function shuffleGrid(grid: number[], size: number, boxRows: number, boxCols: number): void {
  // Shuffle rows within each band
  for (let band = 0; band < boxCols; band++) {
    const startRow = band * boxRows;
    const rowIndices = Array.from({ length: boxRows }, (_, i) => startRow + i);
    const shuffledRows = shuffle(rowIndices);
    swapRows(grid, size, rowIndices, shuffledRows);
  }

  // Shuffle columns within each stack
  for (let stack = 0; stack < boxRows; stack++) {
    const startCol = stack * boxCols;
    const colIndices = Array.from({ length: boxCols }, (_, i) => startCol + i);
    const shuffledCols = shuffle(colIndices);
    swapCols(grid, size, colIndices, shuffledCols);
  }

  // Shuffle the numbers themselves
  const numberMap = shuffle(Array.from({ length: size }, (_, i) => i + 1));
  for (let i = 0; i < grid.length; i++) {
    grid[i] = numberMap[grid[i] - 1];
  }
}

function swapRows(grid: number[], size: number, from: number[], to: number[]): void {
  const temp: number[][] = from.map(row => {
    const rowData: number[] = [];
    for (let col = 0; col < size; col++) {
      rowData.push(grid[row * size + col]);
    }
    return rowData;
  });

  for (let i = 0; i < from.length; i++) {
    for (let col = 0; col < size; col++) {
      grid[to[i] * size + col] = temp[i][col];
    }
  }
}

function swapCols(grid: number[], size: number, from: number[], to: number[]): void {
  const temp: number[][] = from.map(col => {
    const colData: number[] = [];
    for (let row = 0; row < size; row++) {
      colData.push(grid[row * size + col]);
    }
    return colData;
  });

  for (let i = 0; i < from.length; i++) {
    for (let row = 0; row < size; row++) {
      grid[row * size + to[i]] = temp[i][row];
    }
  }
}

// Simple cell removal without unique solution check (fast)
function removeCellsSimple(
  solution: number[],
  targetClues: number
): (number | null)[] {
  const puzzle: (number | null)[] = [...solution];
  const totalCells = solution.length;
  const cellsToRemove = totalCells - targetClues;

  const indices = shuffle(Array.from({ length: totalCells }, (_, i) => i));

  for (let i = 0; i < cellsToRemove && i < indices.length; i++) {
    puzzle[indices[i]] = null;
  }

  return puzzle;
}

// Cell removal with unique solution check (for smaller grids)
function removeCellsWithCheck(
  solution: number[],
  size: GridSize,
  targetClues: number
): (number | null)[] {
  const puzzle: (number | null)[] = [...solution];
  const totalCells = size * size;
  const cellsToRemove = totalCells - targetClues;

  const indices = shuffle(Array.from({ length: totalCells }, (_, i) => i));

  let removed = 0;
  for (const index of indices) {
    if (removed >= cellsToRemove) break;

    const value = puzzle[index];
    puzzle[index] = null;

    if (hasUniqueSolution(puzzle, size)) {
      removed++;
    } else {
      puzzle[index] = value;
    }
  }

  return puzzle;
}

export function generatePuzzleAsync(
  size: GridSize,
  difficulty: Difficulty
): Promise<Puzzle> {
  return new Promise((resolve, reject) => {
    // Use setTimeout to avoid blocking UI
    setTimeout(() => {
      try {
        const puzzle = generatePuzzle(size, difficulty);
        resolve(puzzle);
      } catch (error) {
        reject(error);
      }
    }, 0);
  });
}

// Generate a simple puzzle for testing
export function generateSimplePuzzle(size: GridSize): Puzzle {
  const emptyGrid: (number | null)[] = new Array(size * size).fill(null);
  const result = solveWithRandomization(emptyGrid, size);

  if (!result.solved || !result.solution) {
    throw new Error('Failed to generate puzzle');
  }

  const solution = result.solution;
  const puzzle: (number | null)[] = [...solution];

  // Remove ~40% of cells for a simple puzzle
  const cellsToRemove = Math.floor(size * size * 0.4);
  const indices = shuffle(Array.from({ length: size * size }, (_, i) => i));

  for (let i = 0; i < cellsToRemove; i++) {
    puzzle[indices[i]] = null;
  }

  return {
    id: generateId(),
    size,
    difficulty: 'easy',
    cells: puzzle,
    solution,
  };
}

File: love/src/puzzle/generator.lua
-- 2D puzzle generator.
-- Supports sizes 9 (backtracking + uniqueness) and 16 (fast pattern-based).
-- Returns a Puzzle table: { size, difficulty, cells, solution }
-- cells: 1-indexed flat array, nil = blank, number = given clue.

local Solver = require "src.puzzle.solver"
local V      = require "src.puzzle.validator"

local Gen = {}

-- Clue targets per size and difficulty { min, max }
local CLUE_TARGETS = {
  [4] = {
    easy   = {10, 12},
    medium = { 8, 10},
    hard   = { 6,  8},
    expert = { 4,  6},
  },
  [9] = {
    easy   = {36, 45},
    medium = {32, 36},
    hard   = {28, 32},
    expert = {22, 28},
  },
  [16] = {
    easy   = {130, 160},
    medium = {100, 130},
    hard   = { 80, 100},
    expert = { 60,  80},
  },
}

local function shuffle(t)
  for i = #t, 2, -1 do
    local j = math.random(i)
    t[i], t[j] = t[j], t[i]
  end
  return t
end

local function target_clues(n, difficulty)
  local t = CLUE_TARGETS[n][difficulty]
  return t[1] + math.random(0, t[2] - t[1])
end

-- Remove cells while preserving unique solution (for n ≤ 9).
local function remove_with_check(solution, n, target_clues_count)
  local puzzle = {}
  for i = 1, n * n do puzzle[i] = solution[i] end

  local indices = {}
  for i = 1, n * n do indices[i] = i end
  shuffle(indices)

  local removed = 0
  local to_remove = n * n - target_clues_count

  for _, idx in ipairs(indices) do
    if removed >= to_remove then break end
    local saved = puzzle[idx]
    puzzle[idx] = nil
    if Solver.has_unique_solution(puzzle, n) then
      removed = removed + 1
    else
      puzzle[idx] = saved
    end
  end

  return puzzle
end

-- Simple removal without uniqueness check (fast, for n ≥ 16).
local function remove_simple(solution, target_clues_count)
  local n2 = #solution
  local puzzle = {}
  for i = 1, n2 do puzzle[i] = solution[i] end

  local indices = {}
  for i = 1, n2 do indices[i] = i end
  shuffle(indices)

  local to_remove = n2 - target_clues_count
  for i = 1, math.min(to_remove, n2) do
    puzzle[indices[i]] = nil
  end
  return puzzle
end

-- Pattern-based valid grid for large sizes (16×16).
local function create_pattern_grid(n)
  local bs = V.box_size(n)  -- box rows = box cols = bs
  local grid = {}
  for row = 0, n - 1 do
    for col = 0, n - 1 do
      -- Formula from the TS generator: produces a valid sudoku layout
      local value = (bs * (row % bs) + math.floor(row / bs) + col) % n + 1
      grid[row * n + col + 1] = value
    end
  end
  return grid
end

local function swap_rows(grid, n, from_rows, to_rows)
  local temp = {}
  for i, r in ipairs(from_rows) do
    temp[i] = {}
    for c = 1, n do temp[i][c] = grid[(r - 1) * n + c] end
  end
  for i, r in ipairs(to_rows) do
    for c = 1, n do grid[(r - 1) * n + c] = temp[i][c] end
  end
end

local function swap_cols(grid, n, from_cols, to_cols)
  local temp = {}
  for i, col in ipairs(from_cols) do
    temp[i] = {}
    for r = 1, n do temp[i][r] = grid[(r - 1) * n + col] end
  end
  for i, col in ipairs(to_cols) do
    for r = 1, n do grid[(r - 1) * n + col] = temp[i][r] end
  end
end

local function shuffle_pattern_grid(grid, n)
  local bs = V.box_size(n)

  -- Shuffle rows within each horizontal band
  for band = 0, bs - 1 do
    local start = band * bs + 1
    local rows = {}; for i = 0, bs - 1 do rows[i + 1] = start + i end
    local shuffled = {}; for i, v in ipairs(rows) do shuffled[i] = v end
    shuffle(shuffled)
    swap_rows(grid, n, rows, shuffled)
  end

  -- Shuffle cols within each vertical stack
  for stack = 0, bs - 1 do
    local start = stack * bs + 1
    local cols = {}; for i = 0, bs - 1 do cols[i + 1] = start + i end
    local shuffled = {}; for i, v in ipairs(cols) do shuffled[i] = v end
    shuffle(shuffled)
    swap_cols(grid, n, cols, shuffled)
  end

  -- Remap values
  local map = {}; for v = 1, n do map[v] = v end
  shuffle(map)
  for i = 1, n * n do grid[i] = map[grid[i]] end
end

-- Assess difficulty from clue count and solving simulation.
-- Returns 'easy', 'medium', 'hard', or 'expert'.
local function assess_difficulty(cells, n, target_diff)
  local clues = 0
  for i = 1, n * n do if cells[i] then clues = clues + 1 end end
  -- Simple heuristic: classify by clue density
  local density = clues / (n * n)
  if density >= 0.45 then return "easy"
  elseif density >= 0.37 then return "medium"
  elseif density >= 0.30 then return "hard"
  else return "expert" end
end

-- Public API ──────────────────────────────────────────────────────────────────

-- Generate a puzzle. Returns { size, difficulty, cells, solution }.
-- For n=9: backtracking + unique solution check (may take ~0.5s on ARM).
-- For n=16: fast pattern-based (instant).
function Gen.generate(n, difficulty)
  difficulty = difficulty or "medium"
  local solution, puzzle

  if n >= 16 then
    solution = create_pattern_grid(n)
    shuffle_pattern_grid(solution, n)
    local tc = target_clues(n, difficulty)
    puzzle = remove_simple(solution, tc)
  else
    -- Backtracking generation with unique solution guarantee
    local empty = {}
    for i = 1, n * n do empty[i] = nil end
    solution = Solver.solve_random(empty, n)
    assert(solution, "Failed to generate solution for " .. n .. "×" .. n)
    local tc = target_clues(n, difficulty)
    puzzle = remove_with_check(solution, n, tc)
  end

  local actual_diff = assess_difficulty(puzzle, n, difficulty)

  return {
    size       = n,
    difficulty = actual_diff,
    cells      = puzzle,    -- given clues (nil = blank)
    solution   = solution,  -- complete solution
  }
end

return Gen
